rabbit-escape-render/src/rabbitescape/render/WaterRegionRenderer.java

```diff
@@ -0,0 +1,364 @@
+package rabbitescape.render;
+
+
+import java.util.Vector;
+
+import rabbitescape.engine.BehaviourTools;
+import rabbitescape.engine.Block;
+import rabbitescape.engine.CellularDirection;
+import static rabbitescape.engine.CellularDirection.*;
+import rabbitescape.engine.WaterRegion;
+import rabbitescape.engine.World;
+import rabbitescape.engine.textworld.BlockRenderer;
+import rabbitescape.engine.util.CellDebugPrint;
+import rabbitescape.engine.util.LookupItem2D;
+import rabbitescape.engine.util.Position;
+import rabbitescape.render.gameloop.WaterDynamics;
+
+public class WaterRegionRenderer implements LookupItem2D
+{
+
+    public static final int contentsPerParticle = 50;
+    public static final int maxParticleCountChange = 2;
+    private static final int maxHeightChange = 1;
+
+    public static enum BounceSurface
+    {
+        VERTICAL,
+        HORIZONTAL,
+        LEFT_RISE,
+        RIGHT_RISE
+    }
+
+    public WaterRegion region;
+
+    private int targetWaterHeight = 0;
+    private int height = 0;
+    private int lastHeight = 0;
+
+    public boolean drawnLT, drawnLB, drawnR;
+
+    private final World world;
+    private final WaterDynamics dynamics;
+
+    public WaterRegionRenderer(WaterRegion region, World world, WaterDynamics dynamics )
+    {
+        this.region = region;
+        this.world = world;
+        this.dynamics = dynamics;
+    }
+
+    public void setWaterRegion( WaterRegion region )
+    {
+        this.region = region;
+    }
+
```

Comment:
WaterRegion instances are re-created every game step. WaterRegionRenderer needs persistent state so water levels rise and fall smoothly (and in the future so water droplets move smoothly).

WaterDynamics.step() should use this to connect WaterRegionRenderer with a new WaterRegion. you are correct that it does not. I need to think about this one.